### ISSUE TITLE
help on emotimevideo_cli error:sample size is different 

### ISSUE THREAD

**Issue by hgtunc** [luca-m/emotime#35]
Hello
First of All thanks for your project. I am working on emotimevideo_cli on Ubuntu 14.04 opencv 2.4.9
And I am giving multicolor mp4 video for input but emotimevideo_cli says "Loading 'video.mp4'" and "init done""opengl support avaliable"and then return below error.
Opencv Error: Sizes of input arguments do not match (the sample size is different what has been used for training )in cvPreparePredictData
Aborted core dumped. 
I cannot resolve the problem What do you offer me to do? Is it about video size, color or framerate?

**Comment by luca-m** [comment]
Hi,

May you provide me the command you are running?

These are some commands I used some time ago for demo purposes :

```
#!/bin/bash
## demo_cbcl1_1vsall.sh

echo "--> [S045-002] 1 vs all"
echo 'mmi/S045-002.avi' | ./emotimevideo_cli ../resources/haarcascade_frontalface_cbcl1.xml ../resources/haarcascade_eye.xml 48 48 3 5 4 svm ../models/dataset_svm_354_cbcl1_1vsall/classifiers/svm/*.xml  

echo "--> [S045-002] 1 vs all ext"
echo 'mmi/S045-002.avi' | ./emotimevideo_cli ../resources/haarcascade_frontalface_cbcl1.xml ../resources/haarcascade_eye.xml 48 48 3 5 4 svm ../models/dataset_svm_354_cbcl1_1vsallext/classifiers/svm/*.xml 
```
It's important that parameters such as "48 48 3 5 4" are the same specified in the configuration file used for the training of the models.  These ones works with the models you can download from the release page.

**Comment by hgtunc** [comment]
Hi again,
We change the command as you described and it works.
But we faced another problem. We are using emotimevideo_cli program it predicts emotions correctly. But too slow. 14 seconds video takes 210 seconds to execute. (360 frames executed). We were using virtualized(vmware esxi 5.1 on server) platform switch to another virtualized platform(desktop vmware workstation) that have video card support it does not affect the result.  
Is it same for your enviroment? How we can accelerate the program? Is there any parameter tha we can modify?
Thanks again.
Best regards.

**Comment by luca-m** [comment]
I can run it in realtime with webcam video input, however I think to know what's the matter: it might depend on the resolution of the video. 

I'm going to explain it better.

1) With low resolution video or images the size of the face detected by the face detector should be around N x M pixels (eg. 300x200, its just an estimation, don't take it as ground truth), the extracted face then is used to compute the gabor filter bank, which in a N x M x NWIDTH x NTHETA x NLAMBDA operation and this filtering operation could be performance killing in case of high resolution face because at the moment no parallelism has been implemented.

A possible improvement to avoid this dependency could be to enhance the "src/facedetector/FaceDetector.cpp" face detector adding a resize operation to the detected face in order to define a performance-saving upper bound to that N x M value. 

Eg you could add something like this at the end of the FaceDetector::detect method :

```
Size size(200,300);  // or do some calculation to maintain aspect ratio.. play a bit with it.
resize(src_mat, dst_mat, size);
```

2) The facedetector parameters in "src/facedetector/FaceDetector.cpp" could be sub-optimal. I added a workaround to better configure the haar cascade detector. Eg. at line 28:

```
    if (face_config_file.find(std::string("cbcl1"))!=std::string::npos){
      this->faceMinSize=Size(30, 60);
    } else {
      this->faceMinSize=Size(120,200);
    }
``` 
Maybe these values are not too good for your video, you may want to change it and recompile.

I suggest you to play a bit with these parameters.. Let me know if you think we could need a re-training of the models, in that case I might provide you some support.

Hope it helps!

**Comment by hgtunc** [comment]
Thanks for your reply, resize operation improved response time. Now 15 seconds video takes 80 seconds instead of 210. Still we have to do some improvement, may be doing parallelism for quick response. Do you have any suggestion for parallel processing.  
Also we have another challenge. we did training with Cohan Kanade images by following your training procedure. But finally our trained system result and your linux ready compiled result does not match and compiled version seems true. We dont know what is wrong?
Finally emotime gui stopping time takes 5-10 seconds, prediction finalized but emotime gui still exists 5-10 seconds.
Thanks again.
Best regards.

**Comment by luca-m** [comment]
Hi hgtunc,
Sorry for the late reply but these days were hell days.

15 sec instead of 80 are good, but still not realtime (maybe some other tuning, smaller faces could help) in my realtime webcam session I used a 640x480 webcam (low res) and face was almost 1/4 of the screen (200x160).

Parallelism is a issue we down-prioritized for pragmatical reasons related to initial deadlines, mainly because:

* is not natively supported in the same way in win and linux
* may require a C++ 11 "translation" of the code in order to use new multiplatform threading support
* may require to adopt other libs (eg boost)

But it is a good point because the gabor-bank loop is very easily parallelizable with a fork-join pattern.

Results of complied version does not match the with the old results? I'll check..

**Comment by luca-m** [comment]
give a try with this c9db3944178ee311a597f274bdc895c48198d628

**Comment by luca-m** [comment]
solved with release 1.2
